KAFKA-2389: remove commit type from new consumer.

becketqin opened the pull request:
A shot to remove commit type from new consumer. The coordinator constructor takes a default offset commit callback mainly for testing purpose.

hachikuji commented:
LGTM. I thought it was a little weird before to have a callback in the synchronous case. I've also considered making the behavior explicit by naming the methods "commitSync" and "commitAsync," but I don't think it's necessary.

nehanarkhede commented on clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java:
```diff
@@ -59,24 +59,24 @@
 
     /**
-     * @see KafkaConsumer#commit(CommitType, ConsumerCommitCallback)
+     * @see KafkaConsumer#commit(OffsetCommitCallback)
```
We changed ConsumerRebalanceCallback to ConsumerRebalanceListener. Can we standardize on that naming scheme for callbacks and call this OffsetCommitListener?

ewencp replied:
The intent with that naming was to differentiate between cases where the user callback is invoked once upon completion of an event (commit callbacks should only be invoked once when the action completes) and cases where it may be invoked arbitrarily many times and is not tied to a specific operation (rebalance is triggered externally, so the listener methods may be invoked many times).

However, if that didn't come across, then maybe we haven't done a good enough job naming them...

nehanarkhede replied:
Nuanced. Didn't come across :)

ewencp commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -813,21 +816,18 @@ public void commit(final Map<TopicPartition, Long> offsets, CommitType commitTyp
      */
     @Override
-    public void commit(final Map<TopicPartition, Long> offsets, CommitType commitType, ConsumerCommitCallback callback) {
+    public void commit(final Map<TopicPartition, Long> offsets, OffsetCommitCallback callback) {
```
I realize it's clearly documented in the javadoc, but I am a little concerned about the naming. If you just glance at the options for committing and think "I don't need a callback, I'll just use the version that doesn't include it", you can end up with synchronous commits when what you really wanted was async without a callback, especially if you've only read the javadoc for the version that has the callback.

Although I think it's uglier, I'd suggest naming the one we think will be more commonly used as just `commit` and the other one with sync/async in the name. Then just scanning the APIs makes it clear you need to be careful which one you use. We can do without the renaming, but I suspect we'll end up with lots of mailing list questions on the topic if we leave it as is.

gwenshap replied:
yep, especially given the current producer behavior.

becketqin replied:
@ewencp @gwenshap Good point. Renaming them to commitSync and commitAsync would be more clear.

ewencp commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -840,22 +840,14 @@ public void commit(final Map<TopicPartition, Long> offsets, CommitType commitTyp
-            release();
-        }
+    public void commit(OffsetCommitCallback callback) {
+        commit(subscriptions.allConsumed(), callback);
```
Shouldn't this still have the `acquire`/`release` calls since it invokes `subscriptions.allConsumed()` before calling the other form of `commit`?

ewencp commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Coordinator.java:
```diff
@@ -217,21 +217,6 @@ public void ensureCoordinatorKnown() {
         }
     }
 
-    /**
```
We may have lost valuable information about how async commits can block on the coordinator when this comment was deleted. Not critical, especially since this is internal code, but you might want to restore that comment by moving it to `commitOffsetsAsync`.

becketqin replied:
@ewencp Is this comment true? I did not find blocking code in `commitOffsetAsync`. It looks it just put the OffsetCommitRequest to the NetworkClient and then return the future. Did I miss anything?

hachikuji replied:
I think the comment was old. Previously we blocked when the coordinator wasn't ready, but now we just fail the request to keep it asynchronous.

ewencp replied:
Yeah, my bad, I thought there was an `ensureCoordinatorKnown` call still.

ewencp commented:
Left some comments on things you might want to tweak, but otherwise LGTM.

hachikuji commented:
@becketqin Did the commit history get screwed up somehow? This seems to be including a bunch of unrelated stuff.

guozhangwang commented:
@becketqin could you rebase on trunk? There are too many unrelated diff files which makes it very hard to review.

becketqin committed:
rebased on trunk
Addressed Gwen and Ewen's comments.
Added commitAsync() convenient method as Jay suggested.

becketqin commented:
@guozhangwang @hachikuji Yes.... I must have done something wrong.... I don't know how to fix it. Do you mind if close this PR and open another one?

hachikuji commented:
@becketqin Looks like you may have merged instead of rebasing (there's still the initial commit at the front of the history). You might have to rewind to the initial commit and try the rebase again (maybe someone else knows a better way?). Anyway, if you fix the problem in your local branch, you should be able to do a 'push -f' to update this PR.

becketqin commented:
@hachikuji Awesome, you are right, what I need is a push -f. Thanks for help.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -783,7 +784,13 @@ public void unsubscribe(TopicPartition... partitions) {
         DelayedTask task = new DelayedTask() {
             public void run(long now) {
-                commit(CommitType.ASYNC);
+                commitAsync(new OffsetCommitCallback() {
```
Can we just use DefaultOffsetCommitCallback here?

becketqin replied:
I thought the error message might be different because this one explicitly says "Auto commit".  Would User want to distinguish between auto commit and normal commit? If we don't care about the error message, we may simply call commitAsync() and the DefaultOffsetCommitCallback will be used.

becketqin replied:
I thought about whether we need to create a new callback for each commit, but since auto commit is not frequent and the callback is very small, it might be fine. If you think it matters we can reuse one callback.

hachikuji replied:
I thought about that too. Probably doesn't matter much and it's always kind of nice to have the code inline. As for using DefaultOffsetCommitCallback, either way is ok for me, though I doubt there's much chance of confusion between the auto and manual commits since users would (at least should) probably only be using one of them.

becketqin replied:
I think some application actually will do both - nothing prevents them from doing that. I am slightly preferring having auto commit logged separately for user experience and facilitate debugging, given it is only tiny amount of code.

hachikuji commented:
@becketqin Minor thing, but I think there might be an example in the class-level javadoc for KafkaConsumer that needs updating.

becketqin commented:
@hachikuji Good catch. I missed that one.

becketqin committed:
Addressed Jason's comments.

hachikuji commented:
LGTM

guozhangwang commented:
Thanks @becketqin @hachikuji , I left some comments on the ticket regarding the names. Other than that the implementation LGTM.

becketqin commented:
@guozhangwang Ping for review. I just rebased on trunk. I want to wait for this to be checked in before adding new consumer to mirror maker. Thanks!

becketqin commented:
@guozhangwang ping.

guozhangwang commented:
Hey @becketqin, sorry for being late.. will try to take a look again today.

guozhangwang commented:
LGTM. Thanks for the patch!!
